### PR TITLE
Feat input timeformat

### LINKED ISSUE #10721: input[time] renders with milliseconds if the browser does not support time input
Whenever a browser does not support input[time] natively, angular falls back to text field. It renders nicely in Chrome (e.g. 07:15 AM), but in Firefox it will display 07:15:00.000 instead. There are some quick hacks out there: http://mark.zealey.org/2015/01/08/formatting-time-inputs-nicely-with-angularjs

It would be nice, if we could configure the date format whenever angular switches in fallback mode.

The problem is located in [input.js](https://github.com/angular/angular.js/blob/master/src/ng/directive/input.js#L385) where we set the format to 'HH:mm:ss.sss'

### LINKED ISSUE #16510: input[time] renders with miliseconds in Firefox
**I'm submitting a ...**
- [x ] bug report

**Current behavior:**
Setting value of model to which `input[type="time"]` directive is bounded to, to instance of `Date` makes control display miliseconds in Firefox.

**Expected / new behavior:**
Description at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time documents only optional seconds but not miliseconds. Input field of type `input[type="time"]` should not display miliseconds.

**Minimal reproduction of the problem with instructions:**
Plunker available at https://docs.angularjs.org/api/ng/input/input%5Btime%5D page.

**AngularJS version:** v1.6.10-build.25885+sha.c68b31c (current snapshot)

**Browser:** Firefox 59.0.1

**Anything else:**
This is similar to issue #10721 however not the same as now Firefox does support `input[type="time"]`. Probably both problems stem from Angular not formatting model's value properly when calculating input's value in case of Firefox.

### PR DESCRIPTION
**Pull request by Narretz**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
Browser differences when displaying input time and datetime-local

**What is the new behavior (if this is a feature change)?**
Option to adjust the format

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [x] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:
Closes #10721
Closes #16510 

There are two options:

timeFormatSeconds can be used to configure if seconds or milliseconds should be displayed in general.
timeStripEmptySeconds mimics the behavior of Chrome to hide the seconds completely if they are zero.

### CODE REVIEW

File: src/ng/directive/input.js [hunk 1]
```diff
@@ -255,6 +255,10 @@ var inputType = {
     * The timezone to be used to read/write the `Date` instance in the model can be defined using
     * {@link ng.directive:ngModelOptions ngModelOptions}. By default, this is the timezone of the browser.
     *
+    * The format of the displayed time can be adjusted with the
+    * {@link ng.directive:ngModelOptions#ngModelOptions-arguments ngModelOptions} `timeFormat` and
```
Comment (gkalpak) [inline src/ng/directive/input.js]: `timeFormat` --> `timeSecondsFormat` (?)

File: src/ng/directive/ngModelOptions.js [hunk 2]
```diff
@@ -423,8 +417,72 @@ defaultModelOptions = new ModelOptions({
...
  *
+ * ## Specifying timezones
+ *
+ * You can specify the timezone that date/time input directives expect by providing its name in the
+ * `timezone` property.
+ *
+ *
+ * ## Formatting the value of `time` and `datetime-local`
+ *
+ * With the options `timeFormat` and `timeStripEmptySeconds` it is possible to adjust the value
```
Comment (gkalpak) [inline src/ng/directive/ngModelOptions.js]: `timeFormat` --> `timeSecondsFormat` (?)

File: src/ng/directive/ngModelOptions.js [hunk 3]
```diff
@@ -423,8 +417,72 @@ defaultModelOptions = new ModelOptions({
...
+   <example name="ngModelOptions-time-format" module="timeExample">
+     <file name="index.html">
+       <time-example></time-example>
+     </file>
+     <file name="script.js">
+        angular.module('timeExample', [])
+          .component('timeExample', {
+            templateUrl: 'timeExample.html',
+            controller: function() {
+              this.time = {
```
Comment (gkalpak) [inline src/ng/directive/ngModelOptions.js]: Does this need to be an object?

File: src/ng/directive/ngModelOptions.js [hunk 4]
```diff
@@ -423,8 +417,72 @@ defaultModelOptions = new ModelOptions({
...
+              };
+
+              this.options = {
+                timeSecondsFormat: 'ss',
+                timeStripEmptySeconds: true
+              };
+
+              this.optionChange = function() {
+                this.timeForm.timeFormatted.$overrideModelOptions(this.options);
+                this.time.value = new Date(this.time.value.getTime());
```
Comment (gkalpak) [inline src/ng/directive/ngModelOptions.js]: `getTime()` is redundant.

File: src/ng/directive/ngModelOptions.js [hunk 5]
```diff
@@ -465,6 +527,21 @@ defaultModelOptions = new ModelOptions({
  *     Note that changing the timezone will have no effect on the current date, and is only applied after
  *     the next input / model change.
  *
+ *   - `timeFormat`: Defines if the `time` and `datetime-local` types should show seconds and
```
Comment (gkalpak) [inline src/ng/directive/ngModelOptions.js]: `timeFormat` --> `timeSecondsFormat` (?)

File: src/ng/directive/ngModelOptions.js [hunk 6]
```diff
@@ -465,6 +527,21 @@ defaultModelOptions = new ModelOptions({
...
+ *     milliseconds. The option follows the format string of {@link date date filter}.
+ *     By default, the options is `undefined` which is equal to
+ *     `HH:mm:ss.sss` (shows hours, minutes, seconds and milliseconds). The other options are
+ *     `HH:mm:ss` (strips milliseconds), and `HH:mm`, which strips both seconds and milliseconds.
```
Comment (gkalpak) [inline src/ng/directive/ngModelOptions.js]: This description seems out-of-date.

File: test/ng/directive/inputSpec.js [hunk 7]
```diff
@@ -1384,6 +1384,89 @@ describe('input', function() {
...
+      });
+      expect(inputElm.val()).toBe('1970-01-01T15:41:05');
+
+      ctrl.$overrideModelOptions({timeSecondsFormat: 'ss.sss'});
+
+      $rootScope.$apply(function() {
+        $rootScope.time = new Date(1970, 0, 1, 15, 41, 50, 50);
+      });
+      expect(inputElm.val()).toBe('1970-01-01T15:41:50.050');
+
```
Comment (gkalpak) [inline test/ng/directive/inputSpec.js]: Unnecessary empty line 😱

File: test/ng/directive/inputSpec.js [hunk 8]
```diff
@@ -1644,6 +1727,89 @@ describe('input', function() {
...
+      });
+      expect(inputElm.val()).toBe('15:41:05');
+
+      ctrl.$overrideModelOptions({timeSecondsFormat: 'ss.sss'});
+
+      $rootScope.$apply(function() {
+        $rootScope.time = new Date(1970, 0, 1, 15, 41, 50, 50);
+      });
+      expect(inputElm.val()).toBe('15:41:50.050');
+
```
Comment (gkalpak) [inline test/ng/directive/inputSpec.js]: Unnecessary empty line 😱

**Review by gkalpak — CHANGES_REQUESTED** [review]
Generally LGTM.
I am not a big fun of the names tbh, though:

- `timeFormat`: Not accurate, because the format does not apply to the whole time value.
- `timeSecondsFormat`: Maybe confusing, because when I read "seconds" my mind goes to the two digits after minutes. But arguably, milliseconds are part of the "seconds". So, I can live with that :grin:
- `timeStripEmptySeconds`: Maybe confusing (see above). Also, what does "empty seconds" mean? (I know what it means, but only because I looked at the code :grin:)

Question:
Why not expand formatting to minutes as well (e.g. strip "empty" minutes)? Was it an arbitrary decision? Is it to match the behavior of a popular browser?

**Comment by Narretz** [comment]
Yeah the names aren't great. Since they are already verbose we could change it to "timeStripSecondsIfZero" or maybe "timeStripZeroSeconds".
I think since milliseconds are part of seconds "timeSecondsFormat" is fine. Looks like I did a sloppy job changing it everywhere.

**Comment by Narretz** [comment]
Oh, and why not include minutes: browsers that support time need hours and minutes and will also always display them even if they are 0.

File: src/ng/directive/ngModelOptions.js (hunk 3 above)
Comment (Narretz) [reply src/ng/directive/ngModelOptions.js]: No, but it's good practice ;)

Comment (gkalpak) [reply src/ng/directive/ngModelOptions.js]: <sub>_Disagrees, but not strongly enough to be bothered 😛_</sub>

File: src/ng/directive/input.js [hunk 9]
```diff
@@ -1573,6 +1586,25 @@ function createDateInputType(type, regexp, parseDate, format) {
       }
       return parsedDate;
     }
+
+    function formatter(value, timezone) {
+      var targetFormat = format;
+
+      if (isTimeType && isString(ctrl.$options.getOption('timeSecondsFormat'))) {
+        targetFormat = format.replace('ss.sss', ctrl.$options.getOption('timeSecondsFormat'));
+        if (targetFormat[targetFormat.length - 1] === ':') {
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply src/ng/directive/input.js]: It is ugly and would be simpler with:

```js
targetFormat = targetFormat.replace(/:$/, '');
```

File: src/ng/directive/ngModelOptions.js (hunk 3 above)
Comment (petebacondarwin) [reply src/ng/directive/ngModelOptions.js]: (it is only good practice if you are working directly with the scope - in this case you already have an object on the scope `$ctrl` so it doesn't matter)

**Review by petebacondarwin — APPROVED** [review]
Are you going to rename `timeStripEmptySeconds`?

**Comment by Narretz** [comment]
I want to rename it, but which is better - `stripZeroSeconds` or `stripSecondsIfZero`?

**Comment by petebacondarwin** [comment]
I prefer`stripZeroSeconds`

**Pull request merged**